### PR TITLE
Add a new `BundleIndex` for SourceMap processing

### PR DESCRIPTION
**Pull request by Swatinem**
Instead of querying the Sentry API for files that are needed during SourceMap processing, the goal of the `BundleIndex` is that Sentry proactively feeds a `BundleIndex` to Symbolicator.

This index can then be consulted to determine which exact `ArtifactBundle` a certain file is contained in, or rather which `ArtifactBundle` contains the most recent version of a certain file.

Once the index is fully in place, Symbolicator will not need to query the Sentry API for any files anymore, as the `BundleIndex` fully covers the available files on Sentry. If a file is not in the index, Symbolicator can fall back to scraping in that case.

### CODE REVIEW

File: crates/symbolicator-service/src/services/sourcemap_lookup.rs [hunk 1]
```diff
@@ -713,6 +733,97 @@ impl ArtifactFetcher {
...
+        // as getting to this point means that the index lied to us
+
+        None
+    }
+
+    fn try_get_bundle_from_index(&mut self, key: &FileKey) -> Option<(String, IndexLookupKey)> {
+        // TODO: maybe we want two separate indices for debug_id / url?
+        let bundle_index = self.bundle_index.as_ref()?;
+        let debug_id_index = bundle_index;
+        let url_index = bundle_index;
```
Comment (loewenheim) [inline crates/symbolicator-service/src/services/sourcemap_lookup.rs:801]: Is this just in case we want to use two separate indices later?

**Review by loewenheim — COMMENTED** [review]
Looks good so far. However, I don't quite understand where Symbolicator will be getting the index from.

**Comment by Swatinem** [comment]
> However, I don't quite understand where Symbolicator will be getting the index from.

As the description says, Sentry would proactively push the id/url of the latest index to Symbolicator.
This would also avoid a roundtrip to the artifact-lookup API later on when we are happy with how the index works.
Pushing the latest index from Sentry as part of the symbolicate payload also naturally solves the problem of cache busting.

> Is this just in case we want to use two separate indices later?

Yes, we are still debating if we want to index debug-ids together on a per-release basis, or rather maintain a single per-project debug-id index.

**Comment by loewenheim** [comment]
I'm not sure about the utility of using two indices. They share a type and so both of them have the methods `get_bundle_id_by_url` and `get_bundle_id_by_debug_id`, even though they each only use one of them.

**Comment by Swatinem** [comment]
> I'm not sure about the utility of using two indices. They share a type and so both of them have the methods `get_bundle_id_by_url` and `get_bundle_id_by_debug_id`, even though they each only use one of them.

Agreed, it might be a bit confusing, since it is one type, but can have one or even two sets of data in it.
And it indeed can. You can use the same physical bundle for both the url index and the debug-id index.

I believe this gives us the most flexibility to support the main use-case, but also be open to other use-cases in the future.

The main use-case right now is:
* Have URL-only indexes *per-release*.
* Have DebugId-only indexes *per-project*.

**Review by loewenheim — APPROVED** [review]
Alright, makes sense :)

**Pull request merged**